### PR TITLE
Plugins activation and deactivation

### PR DESCRIPTION
**Pull request by Alroniks**
The task is [BSA-175](https://oat-sa.atlassian.net/browse/BSA-175)

These changes introduce ability to activate and deactivate plugins instead of registering them from scratch.

### CODE REVIEW

File: models/classes/runner/plugins/PluginRegistry.php [hunk 1]
```diff
@@ -30,21 +34,56 @@
...
+    }
+
+    protected function getPluginDefinition(string $pluginId): TestPlugin
+    {
+        if (!$this->isRegistered($pluginId)) {
+            throw new InvalidArgumentException('Requested plugin is not registered');
+        }
+
+        return TestPlugin::fromArray($this->get($pluginId));
     }
```
Comment (wazelin) [inline models/classes/runner/plugins/PluginRegistry.php:88]:
```suggestion
    public function activate(string $pluginId): void
    {
        $this->applyPluginState($pluginId, true);
    }

    public function deactivate(string $pluginId): void
    {
        $this->applyPluginState($pluginId, false);
    }

    protected function getPluginDefinition(string $pluginId): TestPlugin
    {
        if (!$this->isRegistered($pluginId)) {
            throw new InvalidArgumentException('Requested plugin is not registered');
        }

        return TestPlugin::fromArray($this->get($pluginId));
    }

    private function applyPluginState(string $pluginId, bool $isActive): void
    {
        $definition = $this->getPluginDefinition($pluginId);

        $definition->setActive($isActive);

        $this->set($pluginId, $definition->toArray());
    }
```

**Comment by wazelin** [comment]
Could you please check the errors, occuring during the unit tests execution?
Doesn't seem like they were caused by your PR, but it's better to have them fixed anyway. Looks like the `generis.conf` constants are not being defined, but used during the execution.

**Comment by Alroniks** [comment]
@wazelin Could you say which tests exactly, please? Overall TAO or only related to this extension?

**Comment by wazelin** [comment]
> @wazelin Could you say which tests exactly, please? Overall TAO or only related to this extension?

```



1) oat\taoTests\test\unit\runner\plugins\TestPluginServiceTest::testApi

Use of undefined constant CONFIG_PATH - assumed 'CONFIG_PATH' (this will throw an Error in a future version of PHP)



/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/generis/common/ext/class.ConfigDriver.php:42

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/generis/common/oatbox/service/ServiceManager.php:44

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/generis/common/oatbox/AbstractRegistry.php:43

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/taoTests/models/classes/runner/plugins/TestPluginService.php:48

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/taoTests/test/unit/runner/plugins/TestPluginServiceTest.php:71

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/taoTests/test/unit/runner/plugins/TestPluginServiceTest.php:88



2) oat\taoTests\test\unit\runner\plugins\TestPluginServiceTest::testGetAllPlugins

Use of undefined constant CONFIG_PATH - assumed 'CONFIG_PATH' (this will throw an Error in a future version of PHP)



/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/generis/common/ext/class.ConfigDriver.php:42

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/generis/common/oatbox/service/ServiceManager.php:44

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/generis/common/oatbox/AbstractRegistry.php:43

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/taoTests/models/classes/runner/plugins/TestPluginService.php:48

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/taoTests/test/unit/runner/plugins/TestPluginServiceTest.php:71

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/taoTests/test/unit/runner/plugins/TestPluginServiceTest.php:98



3) oat\taoTests\test\unit\runner\plugins\TestPluginServiceTest::testGetOnePlugin

Use of undefined constant CONFIG_PATH - assumed 'CONFIG_PATH' (this will throw an Error in a future version of PHP)



/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/generis/common/ext/class.ConfigDriver.php:42

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/generis/common/oatbox/service/ServiceManager.php:44

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/generis/common/oatbox/AbstractRegistry.php:43

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/taoTests/models/classes/runner/plugins/TestPluginService.php:48

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/taoTests/test/unit/runner/plugins/TestPluginServiceTest.php:71

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/taoTests/test/unit/runner/plugins/TestPluginServiceTest.php:128



4) oat\taoTests\test\unit\runner\providers\TestProviderServiceTest::testApi

Use of undefined constant CONFIG_PATH - assumed 'CONFIG_PATH' (this will throw an Error in a future version of PHP)



/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/generis/common/ext/class.ConfigDriver.php:42

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/generis/common/oatbox/service/ServiceManager.php:44

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/generis/common/oatbox/AbstractRegistry.php:43

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/taoTests/models/classes/runner/providers/TestProviderService.php:43

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/taoTests/test/unit/runner/providers/TestProviderServiceTest.php:70

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/taoTests/test/unit/runner/providers/TestProviderServiceTest.php:87



5) oat\taoTests\test\unit\runner\providers\TestProviderServiceTest::testGetAllProviders

Use of undefined constant CONFIG_PATH - assumed 'CONFIG_PATH' (this will throw an Error in a future version of PHP)



/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/generis/common/ext/class.ConfigDriver.php:42

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/generis/common/oatbox/service/ServiceManager.php:44

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/generis/common/oatbox/AbstractRegistry.php:43

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/taoTests/models/classes/runner/providers/TestProviderService.php:43

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/taoTests/test/unit/runner/providers/TestProviderServiceTest.php:70

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/taoTests/test/unit/runner/providers/TestProviderServiceTest.php:97



6) oat\taoTests\test\unit\runner\providers\TestProviderServiceTest::testGetOneProvider

Use of undefined constant CONFIG_PATH - assumed 'CONFIG_PATH' (this will throw an Error in a future version of PHP)



/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/generis/common/ext/class.ConfigDriver.php:42

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/generis/common/oatbox/service/ServiceManager.php:44

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/generis/common/oatbox/AbstractRegistry.php:43

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/taoTests/models/classes/runner/providers/TestProviderService.php:43

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/taoTests/test/unit/runner/providers/TestProviderServiceTest.php:70

/srv/data/jenkins/workspace/ion-pr_extension-tao-test_PR-336/build/taoTests/test/unit/runner/providers/TestProviderServiceTest.php:124



ERRORS!

Tests: 43, Assertions: 98, Errors: 6.
```

**Comment by Alroniks** [comment]
@wazelin Seems it is work for technical enablers. I will check with them

**Comment by gitromba** [comment]
Configuration is fine, it fails locally too.

**Comment by Alroniks** [comment]
Sorry for your time spent in review of this pull request, but current solution in https://github.com/oat-sa/extension-tao-bosa-selor/pull/36 does not require it anymore. And this pull request should be closed.

**Pull request closed without merging**